pkg/netpol/eval/internal/k8s/policy_connections.go

```diff
@@ -0,0 +1,129 @@
+// connections from previous rules are with higher precedence.
+func (pc *PolicyConnections) UpdateWithRuleConns(ruleConns *common.ConnectionSet, ruleAction string, banpRules bool) error {
+	// banpRules indicates if the rules are coming from BANP; flag used to check the rule Actions are valid since:
+	// Unlike AdminNetworkPolicies that enable: "Pass, Deny or Allow" as the action of each rule.
+	// BaselineAdminNetworkPolicies allows only "Allow and Deny" as the action of each rule.
+	switch ruleAction {
+	case string(apisv1a.AdminNetworkPolicyRuleActionAllow):
+		ruleConns.Subtract(pc.DeniedConns)
+		ruleConns.Subtract(pc.PassConns)
+		pc.AllowedConns.Union(ruleConns)
+	case string(apisv1a.AdminNetworkPolicyRuleActionDeny):
+		ruleConns.Subtract(pc.AllowedConns)
+		ruleConns.Subtract(pc.PassConns)
+		pc.DeniedConns.Union(ruleConns)
+	case string(apisv1a.AdminNetworkPolicyRuleActionPass):
+		if banpRules {
+			return fmt.Errorf(netpolerrors.UnknownRuleActionErr)
+		}
+		ruleConns.Subtract(pc.AllowedConns)
+		ruleConns.Subtract(pc.DeniedConns)
+		pc.PassConns.Union(ruleConns)
+	default:
+		return fmt.Errorf(netpolerrors.UnknownRuleActionErr)
+	}
+	return nil
+}
+
+// CollectANPConns updates the current policyConnections with given conns object from a new admin-network-policy.
+// admin-network-policies are looped by priority order (from lower to higher) , so previous conns take precedence on the conns
+// from the new admin-network-policy
+func (pc *PolicyConnections) CollectANPConns(newAdminPolicyConns *PolicyConnections) {
+	// keep all connections collected from policies with a higher precedence
+	newAdminPolicyConns.DeniedConns.Subtract(pc.AllowedConns)
+	newAdminPolicyConns.DeniedConns.Subtract(pc.PassConns)
+	newAdminPolicyConns.AllowedConns.Subtract(pc.DeniedConns)
+	newAdminPolicyConns.AllowedConns.Subtract(pc.PassConns)
+	newAdminPolicyConns.PassConns.Subtract(pc.DeniedConns)
+	newAdminPolicyConns.PassConns.Subtract(pc.AllowedConns)
+	// add the new conns from current policy to the connections from the policies with higher precedence
+	pc.DeniedConns.Union(newAdminPolicyConns.DeniedConns)
+	pc.AllowedConns.Union(newAdminPolicyConns.AllowedConns)
+	pc.PassConns.Union(newAdminPolicyConns.PassConns)
+}
+
+// CollectConnsFromLowerPrecedencePolicyType updates current PolicyConnections object with connections from a
+// policy with lower priority than ANP. (e.g. network-policy or baseline-admin-network-policy or instead system-default connection)
+// allowed and denied connections of current PolicyConnections object (admin-network-policy) are non-overridden.
+// but pass connections in current PolicyConnections object will be determined by the input PolicyConnections parameter.
+// note that: passConns in otherConns will always be empty. (np and banp don't contain pass connections)
+func (pc *PolicyConnections) CollectConnsFromLowerPrecedencePolicyType(otherConns *PolicyConnections) {
```

Comment:
> * are all `pass` connections expected to be overridden in this function?
yes

1. if the input connection is coming from a netpol then yes; since if it is not allowed by the netpol it is definitely (implicitly) denied.

2. if the input connection is coming from a `banp` : then it may be not mentioned in the banp and then it will be treated as default allow (when returning to the calling func - see [comment](https://github.com/np-guard/netpol-analyzer/pull/380#discussion_r1812039290) 
  
> 
>     * do denied connections include implicit deny from NetworkPolicy resources ?

no and there is no need for that.
if the input connection is coming from an NetworkPolicy.
(current is always from an ANP) 
the flow in this func (as commented and implemented):
1. subtract ANP.allow and ANP.Deny from the NP.allow; so:
what is allowed by the ANP will be kept for sure
what is denied by the ANP will be kept for sure
2. the "remaining" allowed by NP will be union with the allowed by ANP 
here if there are pass by ANP which are allowed by the NP the will be automatically in and allowed 
what is passed by the ANP and not allowed by the NP will not be contained in the allowed 
and in the calling func in `check.go 391-392` the result will only consider the allowed and denied (so every thing is covered) 
--------
please note that this func may not be called twice for same connection;
if the Networkpolicy captures the connection there will be no need to delegate to the BANP .

and all this is explained also in the calling func in `check.go : allAllowedXgressConnections` (see lines 382 - 393 for np capturing the conn with anp) 


>     * what is assumed about input otherConns?
as mentioned in the func docs : otherConns is from a policy in a lower layer than the ANP (either from  NP or from  BANP)
and that passConns in otherConns will always be empty. (np and banp don't contain pass connections)
and more is explained in the calling function in `check.go`

Reply:
> if the input connection is coming from a netpol then yes; since if it is not allowed by the netpol it is definitely (implicitly) denied.

not sure I follow how it is overridden, in case of implicit deny.. the replacement of the `pass` is only for intersection with explicit allow/deny intersections.

Reply:
simplified func by splitting into two (one for np case ; other for banp case)